### ISSUE TITLE
Add support for shared parameters and responses

### ISSUE THREAD

**Issue by pazaan** [stoplightio/studio#4]
**User story.**
As an API designer, I can create shared parameters and responses, so that I don't have to copy and paste shared parameters (like auth headers) and responses (like error responses).

**Is your feature request related to a problem?**
Copy/paste boilerplate is tedious. Especially when you change the definition, which means you have to go and find all the instances across your whole API.

**Describe the solution you'd like**
I'd like Studio to support Shared parameters and responses (or provide a similar mechanism) like Stoplight Next

**Scope & AC**
- "Shared" Responses & Parameters added according to wireframes
- [refactoring] Remove "Embedded Models" from design tree & move the the new tabbed pane

**Out of scope**
- Referencing shared responses and parameters in UI
- Not adding parameters if we don't already support it (e.g. if this would be done before https://github.com/stoplightio/studio/issues/3 it would not include adding "path params")

**Wireframes**

![image.png](https://images.zenhubusercontent.com/5948db0651aa2c3057822fd0/49659b4d-4780-468e-80bc-9a60e111bea1)

**Comment by chris-miaskowski** [comment]
Had a call with @P0lip about it and we discussed potential UX/UI solutions. We got two ideas. I will prepare some wireframes today and ask Marc for approval before starting the implementation.

**Comment by chris-miaskowski** [comment]
@marbemac please take a look and give thumbs up if you agree with the wireframes I put together. It's slightly different than our current approach, curious what you think.

**Comment by chris-miaskowski** [comment]
@P0lip , @mallachari , @lottamus - I would appreciate your feedback too!

**Comment by chris-miaskowski** [comment]
And @philsturgeon 's :)

**Comment by philsturgeon** [comment]
Yeah I would definitely like to see more components be `$ref`'able. We have gone for a simplistic approach to start off with, and some power users will definitely be switching over to the text view for certain sorts of DRYing, but I think we should try to get reusable parameters, requestBodies, responses, and certainly links. Not sure how or when the UI should look, but personally I would like to have this covered somehow.

**Comment by rossmcdonald** [comment]
My 2¢ on the wireframe:

- I like the idea of removing "Embedded Models," I don't think there's a good reason to have them separated out in the sidebar, but I think they should all just be under "Models" and not moved somewhere else.

- I don't like "Standalone Models" and there doesn't seem to be a compelling reason to rename it to that. It should just stay "Models".

- I'm not a fan of adding more form elements to the top-level API screen. Any shared resources should be visible from the sidebar, just like models. Hiding them inside another screen will only make it more difficult to find and use.

**Comment by lottamus** [comment]
> I don't think there's a good reason to have them separated out in the sidebar

@rossmcdonald we used to have standalone and embedded models combine together in the same "Models" folder BUT we separated them as there was confusion around "which one is embedded and which one isn't?". We could solve that with an icon?

> I'm not a fan of adding more form elements to the top-level API screen

Agreed, we could have a "Shared" folder with "Responses", "Parameters", and "Models"?

**Comment by rossmcdonald** [comment]
@lottamus 

> we used to have standalone and embedded models combine together in the same "Models" folder BUT we separated them as there was confusion around "which one is embedded and which one isn't?". We could solve that with an icon?

An icon could work. From the perspective of an API designer, I'm not sure why I care if they are embedded or not as long as I have a way to create a deref'd spec. Who was confused by having them merged? 

> Agreed, we could should have a "Shared" folder with "Responses", "Parameters", and "Models"?

Yeah, having a Shared heading would be much better, though I'm not sure about Models. Wouldn't all models technically be shared?

**Comment by chris-miaskowski** [comment]
Hey @rossmcdonald , @lottamus 

The reason why I think putting it in the tree makes less sense is:
- it will be taking up a lot of space for folks that have 100s APIs
- API-level properties such as `components` will feel detached from the actual file (when, in reality, they aren't)
- Putting stuff in the sidebar makes the UI more complicated. Consider this:
```yaml
parameters:
        - name: petId
          in: path
          required: true
          description: The id of the pet to retrieve
          schema:
            type: string
        - schema:
            type: string
          in: query
          name: q
        - schema:
            type: string
            in: query
            name: h
        - schema:
            type: string
          in: header
          name: h
```

What should the sidebar tree look like?
Should it be (I think it's bloated)
```
- shared
  - parameters
    - q (query)
    - h (header)
    - h (query)
    - petId (path)
```

or (weirdly)

```
- shared
  - parameters
```

or 
```
- shared
  - parameters
    - query
    - path
    - header
```

Also think what the rendered content will be in each of those cases. Our components are very minimalistic. In `next` it made sense to put it on a tree because we actually rendered a lot of stuff in the content area (see below)

![image.png](https://images.zenhubusercontent.com/5948db0651aa2c3057822fd0/793f372b-78af-4b37-b236-e365d0a40e9d)

But with `Studio` it would look a bit funny (the way we render this stuff is much more compact in new app, see below)

![image.png](https://images.zenhubusercontent.com/5948db0651aa2c3057822fd0/a463ba27-7c32-415d-88ae-56d865a77eb3)

That's why I proposed to stick these `components` (and that's a better name imo than `shared` because it's a natural thing in OAS3) in the API view. Similarly to how we render 'methods' (GET, POST, PUT etc)

**Comment by lottamus** [comment]
Personally I like your last suggestion:

```
- shared
  - parameters (17)
    - query (10)
    - path (2)
    - header (5)
  - models (2)
    - todo
    - error
```

That would allow the me to quickly navigate to a UI displaying a list of my shared parameters for the given type.

**Comment by pytlesk4** [comment]
If we are concerned about something being cluttered, why not just remove the parameters folder? 

```
- shared (19)
    - query (10)
    - path (2)
    - header (5)
  - models (2)
    - todo
    - error
```

Would shared be collapsable?

Lastly is it possible to make `PATHS`, `SECURITY`, etc... collapsable? Would help a lot with clutter, hiding it at least.

**Comment by chris-miaskowski** [comment]
@lottamus @pytlesk4 thanks!

So who makes the final call? ;)

**Comment by pytlesk4** [comment]
Is it hard to change? I would suggest testing it out, try it with parameters and without see how it feels?

**Comment by philsturgeon** [comment]
```
- shared
  - parameters (17)
    - query (10)
    - path (2)
    - header (5)
  - models (2)
    - todo
    - error
```

**Comment by chris-miaskowski** [comment]
> Is it hard to change? I would suggest testing it out, try it with parameters and without see how it feels?

Depends on the design we pick but yes that's why I started with a mockup. In general wireframes should be enough for us to decide which one feels better.

Consider this: if I started working on a prototype it would probably take several hours (Studio is tricky). I would present to you folks and get a negative feedback (note that I suggested something completely different that what majority seems to like). Then would have to spend more hours/days to implement/test/debug and the loop would repeat.

Instead, I spent 10 minutes on the wireframes, maybe 15 minutes discussing it and spent the hours I saved by avoiding engineering to finish several other tasks.

My goal is to eventually get wireframes and ACs before we start engineering work (the most costly) across the board for ~all features.

**Comment by chris-miaskowski** [comment]
I think it will become clear who is the owner of this after today's all hands and I will sync with that person.

**Comment by chris-miaskowski** [comment]
Leaving to someone from Studio squad.

**Comment by dgrstl** [comment]
Looks like we put considerable thought into this but suddenly stopped. I see it has been archived in ProductBoard. Assuming this is a precursor to being able to add sharable "things" to the Design Library this feels important. What is our plan?

**Comment by philsturgeon** [comment]
@lottamus it might be worth creating a new issue to replace this one, with whatever the new plan will be. This one got a bit messy.

**Comment by P0lip** [comment]
@philsturgeon @marbemac 
Could we transform the following structure
```
  - Headers (5)
  - URL Params (12)
```
to something reflecting the OAS file, for instance
```
  - parameters (17)
    - query (10)
    - path (2)
    - header (5)
```
`Headers` and `URL Params` are quite tricky from node-management point of view, since they are not nodes in graph, so we'd need a special case to handle them.
I'd really love to avoid it if we can help it.

**Comment by philsturgeon** [comment]
@P0lip I'm trying to strike a balance between generic HTTP stuff and specific OpenAPI terminology. We want to avoid renaming everything, but OAS2 and OAS3 are not even consistent with themselves a lot of the time, and if we add something like RAML, HAR, Postman, etc, then we do not want to be too OASX specific. 

If technically its a giant pain in the butt then do it the way you suggested, just keep in mind the reasoning above with naming stuff in general 🙌

**Comment by marbemac** [comment]
OK I sync'd with jakub, and the tricky bit re splitting out headers, query, path params, etc in the tree as separate items is that those properties don't exist in OpenAPI at all. OpenAPI 2 and 3 jam all parameters into a single `parameters` array that's polymorphic on the `in` property - over the years, from a tooling provider's perspective, this has been the bane of our existence. And now it seems that OpenAPI v3 also allows shared headers in a second spot - the `#/components/headers` property (but still allows `in: header` in `#/components/paremeters` objects, so can have shared headers in two spots now 🙄). Hopefully I'm wrong there?

What we've settled on is probably just a single `Shared Parameters` type tree item that displays a form with a section for each parameter type (query, path, headers, etc). Thoughts?

**Comment by philsturgeon** [comment]
Seems fine to me. Whatever works, we can improve over time.

**Comment by pazaan** [comment]
> And now it seems that OpenAPI v3 also allows shared headers in a second spot - the `#/components/headers` property (but still allows `in: header` in `#/components/paremeters` objects, so can have shared headers in two spots now 🙄). Hopefully I'm wrong there?

Ha! I was just about to come and report a separate bug about this, and came across my original report, which I'd completely forgotten about!
It does indeed appear that headers can be components under both [parameters](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#parameter-object) and [headers](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#header-object) :(
I was coming to report that neither appear to work in the Studio UI currently: I can neither $ref a shared header from `components/headers` or `components/parameters` with an `in: header`. Of course, I can write it in the `code` section, and Spectral is totally fine with it.

**Comment by lottamus** [comment]
@pazaan Unfortunately we still don't support creating shared parameters, though I'm happy to report that we've pushed an update to Studio Web that adds support for $ref'ing them. This feature will land in Studio Desktop v1.10 at the end of the month.

![2020-02-14 13 27 26](https://user-images.githubusercontent.com/7423098/74561315-fb82d700-4f2d-11ea-8ffa-c9306bc8fbf0.gif)

**Comment by woutersmet** [comment]
Sweet! We want to use a variable/parameter for the subdomain in the server (as we distinguish customers/accounts like that), is that covered with this as well?

**Comment by anton-yarchuk** [comment]
Hey,
Many thanks for shared parameters, @lottamus. Works like a charm!

Any updated on shared responses? I'd love to see this feature since many responses (especially errors) are reusable.

**Comment by elipinska** [comment]
I'd love to see shared responses added to Stoplight UI as well, for now I edit them manually in the raw OpenAPI definition but it would be great to have them supported in the form as well.

**Comment by philsturgeon** [comment]
We're tracking all feature requests via the roadmap which is powered by Product Board. Please vote for the feature over here, and you'll get updates when the feature is done. https://roadmap.stoplight.io/c/10-openapi-shared-components

Thank you!

**Comment by philsturgeon** [comment]
This has been done! Studio Web users can enjoy it right now, and Studio Desktop users will get it in the next minor release (should be out soon).

Thanks for your patience folks.
